### ISSUE TITLE
Too many spurious errors/warnings on failing builds with unsupported kernel configs

### ISSUE THREAD

**Issue by udovdh** [openwall/lkrg#70]
```
make[1]: Entering directory '/usr/src/linux-5.11'
  CC [M]  /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.o
In file included from /usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:355,
                 from /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.c:19:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:145:60: warning: ‘struct jump_entry’ declared inside parameter list will not be visible outside of this definition or declaration
  145 | static inline unsigned long p_jump_entry_code(const struct jump_entry *entry) {
      |                                                            ^~~~~~~~~~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h: In function ‘p_jump_entry_code’:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:146:49: error: invalid use of undefined type ‘const struct jump_entry’
  146 |     return (unsigned long)((unsigned long)&entry->code + entry->code);
      |                                                 ^~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:146:63: error: invalid use of undefined type ‘const struct jump_entry’
  146 |     return (unsigned long)((unsigned long)&entry->code + entry->code);
      |                                                               ^~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h: At top level:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:149:62: warning: ‘struct jump_entry’ declared inside parameter list will not be visible outside of this definition or declaration
  149 | static inline unsigned long p_jump_entry_target(const struct jump_entry *entry) {
      |                                                              ^~~~~~~~~~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h: In function ‘p_jump_entry_target’:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:150:49: error: invalid use of undefined type ‘const struct jump_entry’
  150 |     return (unsigned long)((unsigned long)&entry->target) + entry->target;
      |                                                 ^~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:150:66: error: invalid use of undefined type ‘const struct jump_entry’
  150 |     return (unsigned long)((unsigned long)&entry->target) + entry->target;
      |                                                                  ^~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h: At top level:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:153:64: warning: ‘struct jump_entry’ declared inside parameter list will not be visible outside of this definition or declaration
  153 | static inline struct static_key *p_jump_entry_key(const struct jump_entry *entry) {
      |                                                                ^~~~~~~~~~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h: In function ‘p_jump_entry_key’:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:154:24: error: invalid use of undefined type ‘const struct jump_entry’
  154 |     long offset = entry->key & ~3L;
      |                        ^~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../modules/wrap/p_struct_wrap.h:156:55: error: invalid use of undefined type ‘const struct jump_entry’
  156 |     return (struct static_key *)((unsigned long)&entry->key + offset);
      |                                                       ^~
In file included from /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.c:19:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h: At top level:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:364:3: error: #error "LKRG requires CONFIG_KPROBES"
  364 |  #error "LKRG requires CONFIG_KPROBES"
      |   ^~~~~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:378:3: error: #error "LKRG requires CONFIG_JUMP_LABEL"
  378 |  #error "LKRG requires CONFIG_JUMP_LABEL"
      |   ^~~~~
In file included from /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.c:19:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:390:3: error: #error "LKRG does not support RT kernels (PREEMPT_RT is enabled)"
  390 |  #error "LKRG does not support RT kernels (PREEMPT_RT is enabled)"
      |   ^~~~~
make[2]: *** [scripts/Makefile.build:279: /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.o] Error 1
make[1]: *** [Makefile:1800: /usr/src/lkrg-0.9.0] Error 2
make[1]: Leaving directory '/usr/src/linux-5.11'
make: *** [Makefile:96: all] Error 2

**Comment by udovdh** [comment]
Yes RT and such but is this a nice way of handling things?

**Comment by solardiz** [comment]
> Yes RT and such but is this a nice way of handling things?

What do you mean?

**Comment by solardiz** [comment]
> Yes RT and such but is this a nice way of handling things?

Oh, didn't notice right away. So we're correctly refusing to build LKRG with those `#error` lines.

Do you have a better suggestion?

**Comment by udovdh** [comment]
Not really a way but perhaps a little less clutter around the lines.
Not everyone is a programmer and sees through the lines.

**Comment by solardiz** [comment]
Maybe we should try moving the `#error` lines closer to the beginning of `p_lkrg_main.h` and see if this prevents the rest of the error/warning messages from appearing.

Can you try moving these lines:

```c
#if !defined(CONFIG_KPROBES)
 #error "LKRG requires CONFIG_KPROBES"
#elif !defined(CONFIG_HAVE_KRETPROBES)
 #error "CONFIG_KPROBES is enabled, however CONFIG_HAVE_KRETPROBES is not found. LKRG requires both."
#endif

#if !defined(CONFIG_MODULE_UNLOAD)
 #error "LKRG requires CONFIG_MODULE_UNLOAD"
#endif

#if !defined(CONFIG_KALLSYMS_ALL)
 #error "LKRG requires CONFIG_KALLSYMS_ALL"
#endif

#if !defined(CONFIG_JUMP_LABEL)
 #error "LKRG requires CONFIG_JUMP_LABEL"
#endif

#if !defined(CONFIG_STACKTRACE)
/*
 * A #warning in this header file would be printed too many times during build,
 * so let's only do that for something truly important, which the below is not.
 */
// #warning "LKRG does NOT require CONFIG_STACKTRACE. However, in case of pCFI violation, LKRG won't be able to dump full stack-trace."
#endif

#if defined(CONFIG_PREEMPT_RT)
 #error "LKRG does not support RT kernels (PREEMPT_RT is enabled)"
#endif
```

to right after the last `#include` in that file - currently, that would be to line 94.

Does this reduce the clutter?

**Comment by udovdh** [comment]
Yes, in some way:
The error lines appear more towards the top, below it are the compiler  errors.
That is an improvement.

**Comment by solardiz** [comment]
> The error lines appear more towards the top, below it are the compiler errors.
> That is an improvement.

I'm not so sure: it would have e.g. helped me notice the right errors here on GitHub, but when you run `make` in a terminal you probably end up with the last set of errors still on the screen, and need to scroll up for the first error.

**Comment by solardiz** [comment]
Looking at the specific spurious warnings/errors you got, maybe this patch would help (instead of moving those lines):

```diff
+++ b/src/p_lkrg_main.h
@@ -352,7 +352,9 @@ static inline int p_lkrg_counter_lock_val_read(p_lkrg_counter_lock *p_arg) {
 #include "modules/notifiers/p_notifiers.h"                    // Notifiers module
 #include "modules/self-defense/hiding/p_hiding.h"             // Hiding module
 #include "modules/exploit_detection/p_exploit_detection.h"    // Exploit Detection
+#ifdef CONFIG_JUMP_LABEL
 #include "modules/wrap/p_struct_wrap.h"                       // Wrapping module
+#endif
 #include "modules/comm_channel/p_comm_channel.h"              // Communication channel (sysctl) module
 
 

```

**Comment by solardiz** [comment]
Actually, I think this is a better fix:

```diff
+++ b/src/modules/wrap/p_struct_wrap.h
@@ -156,7 +156,7 @@ static inline struct static_key *p_jump_entry_key(const struct jump_entry *entry
     return (struct static_key *)((unsigned long)&entry->key + offset);
 }
 
-#else
+#elif defined(CONFIG_JUMP_LABEL)
 
 static inline unsigned long p_jump_entry_code(const struct jump_entry *entry) {
     return (unsigned long)entry->code;
```

**Comment by solardiz** [comment]
Hmm, looking at the line numbers in @udovdh's build log, apparently `CONFIG_HAVE_ARCH_JUMP_LABEL_RELATIVE` was defined. So it should be something like:

```diff
+++ b/src/modules/wrap/p_struct_wrap.h
@@ -140,7 +140,7 @@ static inline unsigned int p_core_size(struct module *p_mod) {
 
 // #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 20, 0)
 
-#ifdef CONFIG_HAVE_ARCH_JUMP_LABEL_RELATIVE
+#if defined(CONFIG_JUMP_LABEL) && defined(CONFIG_HAVE_ARCH_JUMP_LABEL_RELATIVE)
 
 static inline unsigned long p_jump_entry_code(const struct jump_entry *entry) {
     return (unsigned long)((unsigned long)&entry->code + entry->code);
@@ -156,7 +156,7 @@ static inline struct static_key *p_jump_entry_key(const struct jump_entry *entry
     return (struct static_key *)((unsigned long)&entry->key + offset);
 }
 
-#else
+#elif defined(CONFIG_JUMP_LABEL)
 
 static inline unsigned long p_jump_entry_code(const struct jump_entry *entry) {
     return (unsigned long)entry->code;
```

**Comment by solardiz** [comment]
@udovdh I've just pushed the above fix, but would appreciate you verifying that it helped. Thank you!

**Comment by udovdh** [comment]
@solardiz with your last patch I get:

# make
make -C /lib/modules/5.11.4-rt11/build M=/usr/src/lkrg-0.9.0 modules
make[1]: Entering directory '/usr/src/linux-5.11'
  CC [M]  /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.o
In file included from /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.c:19:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:364:3: error: #error "LKRG requires CONFIG_KPROBES"
  364 |  #error "LKRG requires CONFIG_KPROBES"
      |   ^~~~~
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:378:3: error: #error "LKRG requires CONFIG_JUMP_LABEL"
  378 |  #error "LKRG requires CONFIG_JUMP_LABEL"
      |   ^~~~~
In file included from /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.c:19:
/usr/src/lkrg-0.9.0/src/modules/ksyms/../../p_lkrg_main.h:390:3: error: #error "LKRG does not support RT kernels (PREEMPT_RT is enabled)"
  390 |  #error "LKRG does not support RT kernels (PREEMPT_RT is enabled)"
      |   ^~~~~
make[2]: *** [scripts/Makefile.build:279: /usr/src/lkrg-0.9.0/src/modules/ksyms/p_resolve_ksym.o] Error 1
make[1]: *** [Makefile:1800: /usr/src/lkrg-0.9.0] Error 2
make[1]: Leaving directory '/usr/src/linux-5.11'
make: *** [Makefile:96: all] Error 2

Which is a lot cleaner!

**Comment by solardiz** [comment]
Thank you for testing this, @udovdh!
